From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: anzz1 <anzz1@live.com>
Date: Tue, 30 Jun 2026 17:40:57 +0000
Subject: [PATCH] Spigot API Backports

This patch serves as a consolidation of API backports from Spigot small
enough to not necessitate their own patch

Current backports:

InventoryView.getSlotType()

diff --git a/src/main/java/org/bukkit/inventory/InventoryView.java b/src/main/java/org/bukkit/inventory/InventoryView.java
index c1235f11f9139898265eb39bc8dbed88b5c6e707..8341a7b31cf4baf96925bcf45d0b09fed38eb439 100644
--- a/src/main/java/org/bukkit/inventory/InventoryView.java
+++ b/src/main/java/org/bukkit/inventory/InventoryView.java
@@ -3,6 +3,7 @@ package org.bukkit.inventory;
 import org.bukkit.GameMode;
 import org.bukkit.entity.HumanEntity;
 import org.bukkit.event.inventory.InventoryType;
+import org.bukkit.event.inventory.InventoryType.SlotType;
 
 /**
  * Represents a view linking two inventories and a single player (whose
@@ -208,6 +209,76 @@ public abstract class InventoryView {
         return getTopInventory().getSize() + getBottomInventory().getSize();
     }
 
+    // PandaSpigot start - Spigot API Backports
+    /**
+     * Determine the type of the slot by its raw slot ID.
+     * <p>
+     * If the type of the slot is unknown, then
+     * {@link SlotType#CONTAINER} will be returned.
+     *
+     * @param slot The raw slot ID
+     * @return the slot type
+     */
+    public SlotType getSlotType(int slot) {
+        SlotType type = SlotType.CONTAINER;
+        if (slot >= 0 && slot < getTopInventory().getSize()) {
+            switch (getType()) {
+                case FURNACE:
+                    if (slot == 2) {
+                        type = SlotType.RESULT;
+                    } else if (slot == 1) {
+                        type = SlotType.FUEL;
+                    } else {
+                        type = SlotType.CRAFTING;
+                    }
+                    break;
+                case BREWING:
+                    if (slot == 3) {
+                        type = SlotType.FUEL;
+                    } else {
+                        type = SlotType.CRAFTING;
+                    }
+                    break;
+                case ENCHANTING:
+                case BEACON:
+                    type = SlotType.CRAFTING;
+                    break;
+                case WORKBENCH:
+                case CRAFTING:
+                    if (slot == 0) {
+                        type = SlotType.RESULT;
+                    } else {
+                        type = SlotType.CRAFTING;
+                    }
+                    break;
+                case MERCHANT:
+                case ANVIL:
+                    if (slot == 2) {
+                        type = SlotType.RESULT;
+                    } else {
+                        type = SlotType.CRAFTING;
+                    }
+                    break;
+                default:
+                    // Nothing to do, it's a CONTAINER slot
+            }
+        } else {
+            if (slot == -999) {
+                type = SlotType.OUTSIDE;
+            } else if (getType() == InventoryType.CRAFTING) {
+                if (slot < 9) {
+                    type = SlotType.ARMOR;
+                } else if (slot > 35) {
+                    type = SlotType.QUICKBAR;
+                }
+            } else if (slot >= (countSlots() - 9)) {
+                type = SlotType.QUICKBAR;
+            }
+        }
+        return type;
+    }
+    // PandaSpigot end
+
     /**
      * Sets an extra property of this inventory if supported by that
      * inventory, for example the state of a progress bar.
